### PR TITLE
[Fix] Store dx only for the rows the program owns in layer_norm_bwd_kernel

### PR DESCRIPTION
**Pull request by mark14wu**
## Summary

`layer_norm_bwd_kernel` stores `dx` (and the recomputed `y` / `dres_in`) for its whole `[BT, BD]` tile, but a program only owns `BS = cdiv(T, NS)` rows. When `BT > BS` (small `T` on a GPU with many SMs: `T = 512, G = 2` on 128 SMs gives `BS = 4` with the autotuner's `BT = 32` or `64`), every tile also covers the rows of the next `BT / BS - 1` programs of the group, so up to `BT / BS` programs store the same `dx` rows concurrently. `dw`/`db` are already accumulated over the owned rows only; the three stores were not. The values agree, so results are deterministic, but it is a write-write data race between programs and redundant global traffic. This PR computes the owned-row mask once (`m_o = o_t < min(i_sg * BS + BS, Tg)`) and uses it for the `dx`, `y` and `dres_in` stores and for the `dw`/`db` accumulation (the same expression as before).

Affected: `layer_norm_bwd_kernel` only (the `D > 512` path uses `layer_norm_bwd_kernel1`, unchanged); all dtypes and backends; no numerics change (the rows a program stops storing are stored, with identical values, by the program that owns them). No API change.

## Test plan

- Unit tests added/modified: none (the change is not observable through outputs; see the write-count probe below).
- Dependent tests run: `pytest tests/modules/test_layernorm.py` on this branch: `150 passed, 15 warnings in 276.51s (0:04:36)` (RTX 4090 (sm_89), triton 3.6.0, torch 2.10.0+cu128). The other files listed by `scripts/find_dependent_tests.py fla/modules/layernorm.py` (layer and model tests) were not run locally.
- Varlen / CP / model tests: no.

Reproducer: the script below loads the kernel with its `dx` store rewritten as `tl.atomic_add(p_dx, 1.0, mask=<same mask>)` and launches it with exactly the grid and constexprs `layer_norm_bwd` computes on this GPU, so `dx` counts how many programs stored each element; it then checks the real kernel (forward and backward) against `torch.nn.functional.group_norm` / `layer_norm`.

```python
"""Reproducer for the overlapping dx stores in fla/modules/layernorm.py
(layer_norm_bwd_kernel).

Usage: PYTHONPATH=<flash-linear-attention> python repro_fla_layernorm_bwd.py /path/to/layernorm.py

1. Write-count probe: the kernel's source is loaded with its dx store
   rewritten as tl.atomic_add(p_dx, 1.0, mask=<same mask>), and launched
   with exactly the grid and constexprs layer_norm_bwd() computes for a
   small T on this GPU (NS = min(cdiv(#SM, G), T // G) * G, BS = cdiv(T, NS),
   GS = NS // G, both autotuner tile heights BT = 32 and 64). dx then holds
   the number of programs that stored each element; more than one is a
   write-write race between programs.
2. Correctness: group_norm / layer_norm of the file under test, forward and
   backward, against torch.nn.functional.group_norm.
"""
import importlib.util
import re
import sys
import tempfile

import torch
import triton
import triton.language as tl

from fla.utils import get_multiprocessor_count


def load(path):
    spec = importlib.util.spec_from_file_location("layernorm_under_test", path)
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod


def probe_kernel(src):
    i = src.index("def layer_norm_bwd_kernel(")
    j = src.index("\n@", i)
    body = src[i:j]
    body, n = re.subn(r"tl\.store\(p_dx, b_dx\.to\(p_dx\.dtype\.element_ty\), mask=(\w+)\)",
                      r"tl.atomic_add(p_dx, 1.0, mask=\1)", body)
    assert n == 1, n
    body = body.replace("def layer_norm_bwd_kernel(", "def probe_kernel(")
    tmp = tempfile.NamedTemporaryFile("w", suffix=".py", delete=False)
    tmp.write("import triton\nimport triton.language as tl\n\n\n@triton.jit\n" + body + "\n")
    tmp.close()
    spec = importlib.util.spec_from_file_location("layernorm_probe", tmp.name)
    mod = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(mod)
    return mod.probe_kernel


def main(path):
    mod = load(path)
    probe = probe_kernel(open(path).read())
    torch.manual_seed(0)
    tokens, D, G = 256, 64, 2  # 256 tokens, 2 groups of 64 channels
    T = tokens * G  # layer_norm_bwd views (tokens, G * D) as (tokens * G, D) and passes T = tokens * G
    x = torch.randn(tokens, G * D, device="cuda")
    dy = torch.randn(tokens, G * D, device="cuda")
    w = torch.randn(G * D, device="cuda")
    b = torch.randn(G * D, device="cuda")
    mean = torch.randn(T, device="cuda")
    rstd = torch.rand(T, device="cuda") + 0.5
    NS = min(triton.cdiv(get_multiprocessor_count(x.device.index), G), T // G) * G
    BS = triton.cdiv(T, NS)
    GS = NS // G
    NB = triton.cdiv(T, 2048)
    print(f"tokens={tokens} G={G} D={D} (T={T}) on {get_multiprocessor_count(x.device.index)} SMs: "
          f"NS={NS} programs, BS={BS} rows per program, GS={GS} programs per group")
    for BT in (32, 64):
        dx = torch.zeros(tokens, G * D, device="cuda")
        dw = torch.empty(NS, D, device="cuda")
        db = torch.empty(NS, D, device="cuda")
        probe[(NS,)](x, w, b, None, dy, dx, dw, db, None, None, mean, rstd, T,
                     G=G, D=D, BS=BS, BT=BT, BD=D, NB=NB, GS=GS,
                     IS_RMS_NORM=False, HAS_DRESIDUAL=False, STORE_DRESIDUAL=False,
                     HAS_WEIGHT=True, HAS_BIAS=True, RECOMPUTE_OUTPUT=False)
        writers = dx.view(tokens, G, D)[:, :, 0]  # one count per (token, group)
        print(f"  BT={BT}: programs storing each dx row: min={int(writers.min())} max={int(writers.max())}, "
              f"rows stored by more than one program: {int((writers > 1).sum())}/{writers.numel()}")

    # correctness of the real kernel (forward + backward) against torch
    ok = True
    for (T_, C, G_) in [(256, 128, 2), (256, 64, 1), (4096, 512, 1), (37, 96, 3)]:
        x = torch.randn(T_, C, device="cuda", requires_grad=True)
        w = torch.randn(C, device="cuda", requires_grad=True)
        b = torch.randn(C, device="cuda", requires_grad=True)
        dy = torch.randn(T_, C, device="cuda")
        y = mod.group_norm(x, w, b, num_groups=G_) if G_ > 1 else mod.layer_norm(x, w, b)
        gx, gw, gb = torch.autograd.grad(y, (x, w, b), dy)
        y_ref = torch.nn.functional.group_norm(x, G_, w, b, 1e-5) if G_ > 1 else torch.nn.functional.layer_norm(x, (C,), w, b, 1e-5)
        rx, rw, rb = torch.autograd.grad(y_ref, (x, w, b), dy)
        same = (torch.allclose(y, y_ref, atol=1e-4, rtol=1e-4) and torch.allclose(gx, rx, atol=1e-3, rtol=1e-3)
                and torch.allclose(gw, rw, atol=1e-2, rtol=1e-3) and torch.allclose(gb, rb, atol=1e-2, rtol=1e-3))
        ok &= same
        print(f"  T={T_} C={C} groups={G_}: y/dx/dw/db match torch = {same}")
    print("all correct:", ok)


if __name__ == "__main__":
    main(sys.argv[1])
```

On `main` (RTX 4090 (sm_89), triton 3.6.0, torch 2.10.0+cu128):

```text
tokens=256 G=2 D=64 (T=512) on 128 SMs: NS=128 programs, BS=4 rows per program, GS=64 programs per group
  BT=32: programs storing each dx row: min=1 max=8, rows stored by more than one program: 504/512
  BT=64: programs storing each dx row: min=1 max=16, rows stored by more than one program: 504/512
  T=256 C=128 groups=2: y/dx/dw/db match torch = True
  T=256 C=64 groups=1: y/dx/dw/db match torch = True
  T=4096 C=512 groups=1: y/dx/dw/db match torch = True
  T=37 C=96 groups=3: y/dx/dw/db match torch = True
all correct: True
```

On this branch:

```text
tokens=256 G=2 D=64 (T=512) on 128 SMs: NS=128 programs, BS=4 rows per program, GS=64 programs per group
  BT=32: programs storing each dx row: min=1 max=1, rows stored by more than one program: 0/512
  BT=64: programs storing each dx row: min=1 max=1, rows stored by more than one program: 0/512
  T=256 C=128 groups=2: y/dx/dw/db match torch = True
  T=256 C=64 groups=1: y/dx/dw/db match torch = True
  T=4096 C=512 groups=1: y/dx/dw/db match torch = True
  T=37 C=96 groups=3: y/dx/dw/db match torch = True
all correct: True
```

## Benchmark / NCU (kernel changes only)

- Hardware: RTX 4090, triton 3.6.0, torch 2.10.0+cu128
- Workload: `layer_norm` / `group_norm` backward (`torch.autograd.grad` of `(x, w, b)`), `triton.testing.do_bench(warmup=25, rep=200)`, shapes `(T, D, groups, dtype)` as printed
- Before (`main`):

```text
T=256 D=128 G=2 float32: backward 0.0131 ms
T=512 D=1024 G=1 bfloat16: backward 0.0703 ms
T=2048 D=2048 G=1 bfloat16: backward 0.0847 ms
T=16384 D=2048 G=1 bfloat16: backward 0.2517 ms
T=65536 D=4096 G=1 bfloat16: backward 1.7744 ms
```

- After (this branch):

```text
T=256 D=128 G=2 float32: backward 0.0132 ms
T=512 D=1024 G=1 bfloat16: backward 0.0773 ms
T=2048 D=2048 G=1 bfloat16: backward 0.0778 ms
T=16384 D=2048 G=1 bfloat16: backward 0.2521 ms
T=65536 D=4096 G=1 bfloat16: backward 1.7727 ms
```

- Conclusion: neutral (differences are within run-to-run noise; the removed stores are a small fraction of the kernel's traffic at these shapes). NCU: not run.

## Breaking changes

- None.

## Checklist

- [x] I have read [CONTRIBUTING.md](../CONTRIBUTING.md) and follow its conventions (code style, docstrings, commit prefixes).
- [x] I have read [AGENTS.md](../AGENTS.md) and, where my change matches its scope, the relevant skill under [.agents/skills](../.agents/skills).
- [x] Dependent tests pass locally or in CI, and new behavior is covered by tests where applicable (tick as N/A for changes with no testable code, e.g. docs-only).
- [x] Kernel changes include same-hardware before/after benchmark numbers, dense + varlen where applicable (tick as N/A when no kernel code changed).
- [ ] This PR is minor/cosmetic-only (typo, formatting, style-only tweaks) — tick only if it is, and justify below.

### CODE REVIEW

**Review by zhiyuan1i — APPROVED** [review]
thanks for carefully check
